src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/JenkinsHash.cs

```diff
@@ -0,0 +1,114 @@
+    // Note that we only use the version that takes three ulongs, which was written by the Orleans team.
+    // implementation restored from Orleans v3.7.2: https://github.com/dotnet/orleans/blob/b24e446abfd883f0e4ed614f5267eaa3331548dc/src/Orleans.Core.Abstractions/IDs/JenkinsHash.cs
+    internal static class JenkinsHash
+    {
+        private static void Mix(ref uint a, ref uint b, ref uint c)
+        {
+            a -= b; a -= c; a ^= (c >> 13);
+            b -= c; b -= a; b ^= (a << 8);
+            c -= a; c -= b; c ^= (b >> 13);
+            a -= b; a -= c; a ^= (c >> 12);
+            b -= c; b -= a; b ^= (a << 16);
+            c -= a; c -= b; c ^= (b >> 5);
+            a -= b; a -= c; a ^= (c >> 3);
+            b -= c; b -= a; b ^= (a << 10);
+            c -= a; c -= b; c ^= (b >> 15);
+        }
+
+        // This is the reference implementation of the Jenkins hash.
+        public static uint ComputeHash(byte[] data)
+        {
+            return ComputeHash(data.AsSpan());
+        }
+
+        // This is the reference implementation of the Jenkins hash.
+        public static uint ComputeHash(ReadOnlySpan<byte> data)
+        {
+            int len = data.Length;
+            uint a = 0x9e3779b9;
+            uint b = a;
+            uint c = 0;
+            int i = 0;
+
+            while (i + 12 <= len)
+            {
+                a += (uint)data[i++] |
+                    ((uint)data[i++] << 8) |
+                    ((uint)data[i++] << 16) |
+                    ((uint)data[i++] << 24);
+                b += (uint)data[i++] |
+                    ((uint)data[i++] << 8) |
+                    ((uint)data[i++] << 16) |
+                    ((uint)data[i++] << 24);
+                c += (uint)data[i++] |
+                    ((uint)data[i++] << 8) |
+                    ((uint)data[i++] << 16) |
+                    ((uint)data[i++] << 24);
+                Mix(ref a, ref b, ref c);
+            }
+            c += (uint)len;
+            if (i < len)
+                a += data[i++];
+            if (i < len)
+                a += (uint)data[i++] << 8;
+            if (i < len)
+                a += (uint)data[i++] << 16;
+            if (i < len)
+                a += (uint)data[i++] << 24;
+            if (i < len)
+                b += (uint)data[i++];
+            if (i < len)
+                b += (uint)data[i++] << 8;
+            if (i < len)
+                b += (uint)data[i++] << 16;
+            if (i < len)
+                b += (uint)data[i++] << 24;
+            if (i < len)
+                c += (uint)data[i++] << 8;
+            if (i < len)
+                c += (uint)data[i++] << 16;
+            if (i < len)
+                c += (uint)data[i++] << 24;
+            Mix(ref a, ref b, ref c);
+            return c;
+        }
+
+        public static uint ComputeHash(string data)
+        {
+            byte[] bytesToHash = Encoding.UTF8.GetBytes(data);
+            return ComputeHash(bytesToHash);
```

Comment:
The intermediate allocation for the bye-array can be avoided.

```suggestion
            int maxByteCount = Encoding.UTF8.GetMaxByteCount(data.Length);

            Span<byte> buffer = maxByteCount <= 256
                ? stackalloc byte[256]
                : new byte[maxByteCount];

            int written = Encoding.UTF8.GetBytes(data, buffer);

            return ComputeHash(buffer.Slice(0, written));
```
(assuming `data.Length` will most likely be <= 256, otherwise the byte-array fallback could be rented from the array pool instead)

Is it worth to complicate the code here?

Reply:
removed all `JenkinsHash` methods except `ComputeHash(ReadOnlySpan<byte>)` (not used)